### PR TITLE
fix(protocol-designer): fix dismissing warnings for unsaved step forms

### LINKED ISSUE #6263: PD Bug: cannot dismiss form-level warnings for never-saved step
## Overview

Form-level warnings can be dismissed for saved steps, but for new steps that haven't been saved yet, they can't be.

## Steps to reproduce

- Create a Transfer step with a P1000 pipette
- Set Volume Per Well to 1
- Notice there's a form-level warning: "WARNING Specified volume is below pipette minimum"
- Click the "x" to dismiss the warning
- Nothing happens (console logs: "Tried to dismiss form warning with no stepId")
- Save the step and re-open it. Then, the warning will be dismissable

## Acceptance criteria

Test coverage

### PR DESCRIPTION
**Pull request by IanLondon**
# Overview

Closes #6263

# Changelog

- Add tests to dismiss reducers
- Fix bug where form/timeline warnings could not be dismissed for unsaved forms

# Review requests

Vocab reminder: "Presaved" = new form that has never been saved :upside_down_face: 

- [ ] Timeline and form warnings (eg not enough liquid in well, below minimum volume respectively) should be dismissable for the presaved forms.
- [ ] No weird edge cases introduced in saving and loading protocols trying to leave the presaved form still open when you save the file (`dismiss` does keep metadata in the protocol file, and could save the presaved form dismiss state...)
- [ ] Actually-saved forms should remember that their form and timeline warnings have been dismissed both when you load them in the same session of PD, and also if you dismiss, save the file, reload PD, load the file, and open up the form

# Dev notes

This is a little weird bc it uses a new pseudo-id. Though there's a step `id` initially created in the `ADD_STEP` action, and we can look at it at `state.stepForms.unsavedForm?.id`, there's a problem. The selectors that we have are written to not really consider the presaved form as a step but rather as a terminal item; eg `getSelectedStepId` selector will be `null` because 1. the presaved step is not in `orderedStepIds` and 2. `isStep: false` in `state.ui.steps.selectedItem`.

There is an alternative approach to this PR where we could use the actual presaved step id as the dismiss id, but we'd have to have a special selector that looks into a combo of `unsavedForm` (to get the ID) and `state.ui.steps.selectedItem` (to figure out if it's a presaved step or not -- it could also look in orderedStepIds or something to figure out if it's a presaved step or not). Then the dismiss reducer could use that ID, and knowing whether it's presaved vs legit saved, it could handle the `CANCEL_STEP_FORM` action correctly (which should remove both form + timeline occurrences of the id from the dismiss state, only if that id is presaved, vs do nothing if that id is legit saved)

But instead of messing with the transient ID of presaved forms, I made a new special ID that's only for `dismiss` (a mini-alternative would be reusing `export const PRESAVED_STEP_ID: '__presaved_step__' = '__presaved_step__'` from `steplist/types.js`, but I figured it's already weird enough dealing with the presaved step without canonizing its pseudo-id across concerns)

I think that the existing timeline selection UX benefits from looking at a presaved step as a special terminal item, and that's why we oriented the selectors that way. We're considering big changes to the timeline UX in the future though, so IDK if that will still be true -- if it's not, maybe we could simplify this. Either way we might want to qualify steps better, eg `isSavedStep: boolean` instead of `isStep`??

Unlike the `step-forms` and `ui/steps` modules which are pretty core to PD, `dismiss` is a pretty simple and peripheral module so it kinda makes sense for it to model presaved steps in its own way instead of modifying `step-forms` and `ui/steps` to tread presaved steps as `isStep: true` instead of `isStep: false`.

If you can think of a nicer solution please share it!

# Risk assessment

Low to medium, PD only. Maybe could whitescreen or something

### CODE REVIEW

**Review by shlokamin — COMMENTED** [review]
I'm hesitant about having two ids refer to the same step depending on the scenario. I'm worried this will be a difficult thing to reason with, and another thing to remember to maintain/think about. While it might be a bit annoying to deal with the selectors, I think using the actual presaved step id as the dismiss id is a better option. Down to have a bigger discussion about this though because you've got more context than I do, and I'm also cool with merging this for now and coming back to it later since at the end of the day it's fixing a bug

File: protocol-designer/src/dismiss/selectors.js [hunk 1]
```diff
@@ -33,8 +34,13 @@ export const getDismissedFormWarningTypesForSelectedStep: Selector<
 > = createSelector(
   getDismissedFormWarningTypesPerStep,
   getSelectedStepId,
-  (dismissedWarnings, stepId) =>
-    (stepId != null && dismissedWarnings[stepId]) || []
+  (dismissedWarnings, stepId) => {
+    console.log({ stepId, dismissedWarnings })
```
Comment (shlokamin) [inline protocol-designer/src/dismiss/selectors.js:38]: leftover console.log

File: protocol-designer/src/dismiss/reducers.js [hunk 2]
```diff
@@ -77,6 +73,13 @@ const dismissedWarnings: Reducer<DismissedWarningState, any> = handleActions(
       action: LoadFileAction
     ): DismissedWarningState =>
       getPDMetadata(action.payload.file).dismissedWarnings,
+    CANCEL_STEP_FORM: (
+      state: DismissedWarningState,
+      action: CancelStepFormAction
+    ): DismissedWarningState => ({
+      form: omit(state.form, PRESAVED_STEP_FORM_PSUDO_ID),
+      timeline: omit(state.timeline, PRESAVED_STEP_FORM_PSUDO_ID),
+    }),
```
[the remark this answers is not in the document]
Comment (IanLondon) [reply protocol-designer/src/dismiss/reducers.js:82]: it's just cleaning up the dismiss state so that it doesn't remember stale IDs. Dismiss state is saved in the protocol, so over time you might load a protocol, add a step, dismiss the warnings, cancel that step without saving the step, save the protocol, and repeat. This would make your dismiss state full of stale UUIDs over time. Not a big issue just keeping it tidy

[the remark this answers is not in the document]
Comment (shlokamin) [reply protocol-designer/src/dismiss/reducers.js]: ahh makes sense

**Comment by IanLondon** [comment]
@shlokamin 

> it might be a bit annoying to deal with the selectors

I think it'd require making a new selector that should only be used for dismiss id, which puts the presaved step form on the same level as the saved step forms. This new selector would contrast against the pattern for all the other selectors, where the presaved step isn't included with the rest of the steps. My worry about it is that if somebody later uses that new selector for something besides dismiss, it gets hard to figure out if "steps" includes the presaved one or not (whereas now, it doesn't, presaved is a pseudo-step aka a terminal item)

Maybe the long-term solution is to not assign presaved steps an id until they're saved? I'd want to confirm that makes sense by tracing it out from ADD_STEP onwards. It does seem kinda wrong to give it a UUID so early because since we're treating a presaved step as "terminal item"/pseudostep and the other terminal items don't have UUIDs.

Ie this stuff in `steplist/types.js` hints that the presaved step should always have a pseudo-id, not a UUID

```js
export const START_TERMINAL_ITEM_ID: '__initial_setup__' = '__initial_setup__'
export const END_TERMINAL_ITEM_ID: '__end__' = '__end__'
export const PRESAVED_STEP_ID: '__presaved_step__' = '__presaved_step__'
export type TerminalItemId =
  | typeof START_TERMINAL_ITEM_ID
  | typeof END_TERMINAL_ITEM_ID
  | typeof PRESAVED_STEP_ID
```

File: protocol-designer/src/dismiss/constants.js [hunk 3]
```diff
@@ -0,0 +1,5 @@
+// @flow
+import { PRESAVED_STEP_ID } from '../steplist/types'
+
+export const getStepIdOrPseudoId = (stepId: ?string): string =>
+  stepId || PRESAVED_STEP_ID
```
Comment (shlokamin) [inline protocol-designer/src/dismiss/constants.js:5]: with #6533 is `getStepIdOrPseudoId` still necessary?

Comment (IanLondon) [reply protocol-designer/src/dismiss/constants.js]: Nope! Good catch

**Comment by shlokamin** [comment]
Copying [our slack convo](https://opentrons.slack.com/archives/DQH8BC55W/p1600116092011300) over here for posterity. I had a question about using `getNonNullSelectedItem`, though this is not technically introduced in this PR, because `getSelectedStepId` actually uses `getNonNullSelectedItem`:

> actually, now that I look at what getNonNullSelectedItem is doing, it sometimes does this: 
> `return { isStep: true, id: last(orderedStepIds) }`
> 
> is it okay to dismiss the warning of something with the id of the last ordered step?
> 
> or would that be an unintended side effect?
> 
> it seems like we should always be dispatching the dismissTimelineWarning action with either state.selectedItem.id , or doing nothing and doing a console.warn or something
> 
> (in the event that state.selectedItem.id is null)

TLDR; my concern is that we'll be dispatching the dismiss warning action with a step id that we actually no not intend to.

**Comment by IanLondon** [comment]
@shlokamin I renamed `getNonNullSelectedItem` to `getSelectedItem` when I realized why that selector exists -- it's not because null is bad due to Flow stuff, but it's because when you delete a selected step, the next selected item is the last step in the list (or if there's no steps,. But the `selectedItem` reducer has no way to know bc it can't access `stepForms.orderedSteps` in the state. (Alternatively, we could add `action.[payload|meta].orderedSteps` to the DELETE_STEP action but that feels grosser?)

So `getSelectedItem` isn't a weird artifact of avoiding null, rather it's the source of truth about what item is selected -- `null` is never the case. Nobody reads `ui.steps.selectedItem` directly, but always go through that selector.

LMK what you think!

The still-existing flaw is that if for some reason we showed dismissable warnings when "Starting Deck State" or "Final Deck State" was selected, then the dismiss warning actions would reference those ids and `dismiss` state would remember that dismissal with those IDs. But it doesn't make sense why/how we would show the warnings for some step when no step is selected, so I don't think bad. If we wanted to be really strict we could ignore those special ids from the dismissal state or the dismiss actions.

**Comment by shlokamin** [comment]
Ah, I think I understand now... I wonder if it would be worth naming `ui.steps.selectedItem` to something else so that the concept of "selected step item" and `ui.steps.selectedItem` don't get conflated. 

Though, I guess that's the whole point of selectors. They provide an interface for which consumers access the store, we shouldn't be necessarily be concerned about store structure/naming.`getSelectedItem` is quite semantic, it tells us exactly what the value represents.

File: protocol-designer/src/components/StepEditForm/FormAlerts.js [hunk 4]
```diff
@@ -6,7 +6,7 @@ import {
   actions as dismissActions,
   selectors as dismissSelectors,
 } from '../../dismiss'
-import { getSelectedStepId } from '../../ui/steps'
+import { getSelectedItem } from '../../ui/steps/selectors'
```
Comment (shlokamin) [inline protocol-designer/src/components/StepEditForm/FormAlerts.js:9]: should this be imported from `index.js` instead?

**Review by shlokamin — APPROVED** [review]
Tests look thorough and code looks good, thanks for this @IanLondon!

File: protocol-designer/src/components/StepEditForm/FormAlerts.js (hunk 4 above)
Comment (IanLondon) [reply protocol-designer/src/components/StepEditForm/FormAlerts.js:9]: I'be been trying to gradually deprecate `index.js` usage because it has been the source of circular dependency problems especially around reducers + selectors + thunks
